store/tikv/2pc.go

```diff
@@ -951,29 +952,45 @@ func (c *twoPhaseCommitter) execute(ctx context.Context) (err error) {
 		logutil.SetTag(ctx, "commitTs", commitTS)
 	}
 
-	tryAmend := c.isPessimistic && c.connID > 0 && !c.isAsyncCommit() && c.txn.schemaAmender != nil
-	if !tryAmend {
-		_, _, err = c.checkSchemaValid(ctx, commitTS, c.txn.txnInfoSchema, false)
+	if c.connID > 0 {
+		failpoint.Inject("beforeSchemaCheck", func() {
+			failpoint.Return()
+		})
+	}
+
+	if c.isAsyncCommit() {
+		schemaVerIsTheSame, err := checkSchemaVersionForAsyncCommit(ctx, c.startTS, commitTS, c.store)
```

Comment:
I think we should define a new error variable here and set `err` to that. Then, we can clean up the locks in the defer function.

Reply:
Do you mean we should return an error with a type like  `ErrTxnShouldRollback` and if the error type is it the 2pc cleanup will be done?  Or something like `ErrSchemaHasChanged` for the `checkSchemaVersionForAsyncCommit`?

Reply:
Maybe not necessary. In the defer function we will cleanup when `err != nil`. So if we want to clean up, I think we just make sure `err` is set.

Code like `schemaVerIsTheSame, err := ...` will create a new `err` variable instead of modifying the existing one because of `:=`. Then, the defer function won't be aware of the error.

Reply:
For some of the cleanup cases, the return statement is used to set the error value. Maybe we need to do some refactoring on this to make them look similar.

Reply:
Ah, I see. No problem here.